### PR TITLE
Updated uflow

### PR DESCRIPTION
**Pull request by mtailanian**
## 📝 Description

- Provide a clear summary of the changes and the issue that has been addressed.
- 🛠️ Fixes # (issue number)

fixed loss computation

## ✨ Changes

- added new anomaly segmentation method, from the uflow paper
- updated readme with final results, images, and paper link and citation

Select what type of change your PR is:

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [x] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [x] 📚 I have made the necessary updates to the documentation (if applicable).
- [x] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

### CODE REVIEW

**Comment by mtailanian** [comment]
Hello guys, I just wanted to ask: am I supposed to add something or correct something? should I update the branch?

File: src/anomalib/models/image/uflow/nfa_tree.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
+import itertools as it
```
Comment (ashwinvaidya17) [inline src/anomalib/models/image/uflow/nfa_tree.py:1]: Can you add doc string and a license header to this file?

File: src/anomalib/models/image/uflow/nfa_tree.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+import networkx as nx
+import numpy as np
+import torch
+from mpmath import mp
+from skimage.morphology import max_tree
+
+mp.dps = 15
+
+
+class NFATree:
```
Comment (ashwinvaidya17) [inline src/anomalib/models/image/uflow/nfa_tree.py:12]: Can you add typing and docstring to the methods of this class?

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
Thanks for updating this. I have a few minor comments.

File: src/anomalib/models/image/uflow/nfa_tree.py (hunk 2 above)
Comment (mtailanian) [reply src/anomalib/models/image/uflow/nfa_tree.py:30]: Also added. Thanks!

**Comment by mtailanian** [comment]
I guess it should be fine now... Should I update the branch in order to merge this?

---

Btw, I have one more question.
As you might have seen in the code, I included a method for finding an automatic segmentation of the anomalies, as this method is part of the paper. But it is unused in the current code. 

Is there a good way to include it as part of the default outputs? 
I saw that all the segmentations are actually obtained by maximizing some metric in testing sets, so I'm not sure if it's a good idea to try to add this method there... I guess not, but I can't think of any other place.

Or maybe it is just fine as it is... The code is there in case anyone wants to use it...

Any thoughts are appreciated.
Thanks!

**Comment by mtailanian** [comment]
Hi guys, hope everything is well.
Is there anything I could do to get this merged?
Thanks again!

**Comment by ashwinvaidya17** [comment]
@mtailanian the PR looks good to me. Do you have access to the logs here: https://github.com/openvinotoolkit/anomalib/actions/runs/9483018344/job/26129442354?pr=2104? If so, can you fix the pre-merge checks. Ideally, you should be able to run these locally as well. Just do `pip install pre_commit` and then run `pre-commit run --all-files`

**Comment by ashwinvaidya17** [comment]
> I guess it should be fine now... Should I update the branch in order to merge this?
> 
> Btw, I have one more question. As you might have seen in the code, I included a method for finding an automatic segmentation of the anomalies, as this method is part of the paper. But it is unused in the current code.
> 
> Is there a good way to include it as part of the default outputs? I saw that all the segmentations are actually obtained by maximizing some metric in testing sets, so I'm not sure if it's a good idea to try to add this method there... I guess not, but I can't think of any other place.
> 
> Or maybe it is just fine as it is... The code is there in case anyone wants to use it...
> 
> Any thoughts are appreciated. Thanks!

Maybe you can create an issue for this, and we can refactor this so that users might be able to switch between both. I am not sure what's the priority for this, but it will remain visible until when we finally get around to doing this.

**Comment by mtailanian-digitalsense** [comment]
> @mtailanian the PR looks good to me. Do you have access to the logs here: https://github.com/openvinotoolkit/anomalib/actions/runs/9483018344/job/26129442354?pr=2104? If so, can you fix the pre-merge checks. Ideally, you should be able to run these locally as well. Just do `pip install pre_commit` and then run `pre-commit run --all-files`

Hi @ashwinvaidya17 , thank you for your response!
I had already run the pre-commit. However, I'm a bit confused. pre-commit suggest to change the typing, for example `list -> List`, or `tuple -> Tuple`. I changed it, but the pre-commit itself would revert it to list, tuple, etc...

How should I do? I'm kind of stuck...

Thanks again!

**Comment by mtailanian** [comment]
> @mtailanian the PR looks good to me. Do you have access to the logs here: https://github.com/openvinotoolkit/anomalib/actions/runs/9483018344/job/26129442354?pr=2104? If so, can you fix the pre-merge checks. Ideally, you should be able to run these locally as well. Just do `pip install pre_commit` and then run `pre-commit run --all-files`

Hi @ashwinvaidya17 , thank you for your response!
I had already run the pre-commit. However, I'm a bit confused. pre-commit suggest to change the typing, for example `list -> List`, or `tuple -> Tuple`. I changed it, but the pre-commit itself would revert it to list, tuple, etc...

How should I do? I'm kind of stuck...

Thanks again!

File: docs/source/reference_guide/algorithms/uflow.rst
Comment (samet-akcay) [inline docs/source/reference_guide/algorithms/uflow.rst:1]: Content-wise, is this file different than https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/reference/models/image/uflow.md? With v1, we switched to markdown from rst. If there is not much difference in the content, maybe we could remove this one?

File: src/anomalib/models/image/uflow/nfa_tree.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
```
Comment (samet-akcay) [inline src/anomalib/models/image/uflow/nfa_tree.py:30]: Would it be possible to port the np computations to torch in this implementation? We try to minimise the use of numpy and use torch instead.

**Review by samet-akcay — COMMENTED** [review]
@mtailanian thanks for your patience. It took me quite some time to review this PR. I have two comments below

File: docs/source/reference_guide/algorithms/uflow.rst
Comment (mtailanian) [reply docs/source/reference_guide/algorithms/uflow.rst:1]: Yep, it is identical. I just deleted the .rst

**Comment by samet-akcay** [comment]
> Hi @samet-akcay , thanks for your feedback. I just addressed both comments! Let me know if there is something else... Best,

Thanks @mtailanian

**Comment by mtailanian** [comment]
@samet-akcay I see I have again the same pre-commit errors. Please see [this comment](https://github.com/openvinotoolkit/anomalib/pull/2104/#issuecomment-2203055081). I don't really know how to solve it...

**Comment by samet-akcay** [comment]
It's alright, I'll merge it to a branch, from which I'll try to fix it

**Comment by mtailanian** [comment]
Thanks so much @samet-akcay !
:)
